### PR TITLE
Clarify what no_std is good for.

### LINKED ISSUE #23057: ICE when save a closure to static mut produced from closure call
code sample:
```rust
static mut add2: Box<FnMut() -> i32> = Box::new((||{let mut a = 0; move || {a+=1;a}})());
```

compiler complans:
```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: fcx.inh.deferred_call_resolutions.borrow().is_empty()', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc_typeck/check/vtable.rs:268

```

[playpen](http://is.gd/mb7JuU)

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #23057 

I'd be happy to modify it if we want to say something else.

### CODE REVIEW

File: src/doc/trpl/unsafe.md [hunk 1]
```diff
@@ -426,9 +426,13 @@ Current valid options are:
 
 # Avoiding the standard library
 
-By default, `std` is linked to every Rust crate. In some contexts,
-this is undesirable, and can be avoided with the `#![no_std]`
-attribute attached to the crate.
+By default, `std` is linked to every Rust crate. In some contexts, such as
+writing a kernel, this is undesirable. Rust's standard library can still be
+used when writing things like shared libraries, even when being called into
+like C.
```
Comment (tshepang) [inline src/doc/trpl/unsafe.md]: Did you mean "called into, like C"?

Comment (steveklabnik) [inline src/doc/trpl/unsafe.md]: yes

Comment (tshepang) [inline src/doc/trpl/unsafe.md]: Would be clear to add a comma there, or change that bit to:

> called into (as in C)

Comment (steveklabnik) [inline src/doc/trpl/unsafe.md]: sure thing

**Comment by alexcrichton** [comment]
I think there's a typo in the bug number? Did you mean #20537?

I'm also not sure that this is sufficient to close that issue, I think that we ourselves haven't quite figured out the `no_std` story quite just yet (hence it being an unstable feature). Also, quite a bit of time has passed since its opening, and the landscape has changed!

I think that the bug isn't really a documentation bug but rather one where we need to discuss how to stabilize the feature, "fixing" the bug in the process. Basically I think that the bug should be interpreted as "it would be nice for `#[no_std]` to be stable and all that entails".

**Comment by steveklabnik** [comment]
It is a typo, whoops :(

> I think that the bug isn't really a documentation bug but rather one where we need to discuss how to stabilize the feature

Sounds fine to me. I'll move the original issue over to RFCs.

**Pull request closed without merging**